### ISSUE TITLE
Reconsider addons API

### ISSUE THREAD

**Issue by cortesi** [mitmproxy/mitmproxy#2144]
This is a discussion ticket to get input on our next steps with the addons API. This is a critical piece of our user-facing surface area, and we're very keen to have it be as fluid and usable as possible. The work done in #2129 has moved things to the point where a reassessment is in order.

I'll outline the issues with the API as it stands here, and then I'll put my proposal for a refactoring in a comment on the ticket. If you have ideas, please respond below - I'll update the list of issues we should be thinking about as we go. 

- The startup event sequence for addons is quite confusing - we have ```start()``` -> ```configure()``` -> ```running()```, all of which are called in different contexts, and have names that are close to each other semantically.
- The ```start()``` method takes an options argument, but addons can only create new options at this point, not read and use option values. This is confusing.
- The ```start()``` method can replace the current addon with a new addon by returning it. I feel this is too implicit, and isn't flexible enough. We also may want to expand what addons can do in this event later, and returning a value boxes us in.
- When we first designed this API, we felt that importing a context object (ctx) to do logging or access master was a good light-weight pattern, partly because this is similar to what flask does. It's time to think about this again, even if we decide to keep things as they are. 
- The addons API is extremely lightweight at the moment - just a sequence of events. This has many advantages, and some disadvantages. There are many common patterns that we could support better by having addons derive from a common class - for instance, enabling/disabling addons is very common, and has to be hand-coded every time at the moment. We could also have addons pre-declare the options that they are interested in, and then call configure() only when something relevant to the addon happens (as opposed to calling all configure() methods for all updates). 
- We need a custom events mechanism to allow addons to declare and handle one-off events. One example of this is triggering reading flows from a file. Doing this by having an addon handle an option change or retrieving and explicitly invoking a method on an addon is clumsy.

**Comment by cortesi** [comment]
After pondering this, my proposal is very minimal. On balance, I think we should keep the ctx import rather than making it a parameter to events, and I don't think that moving to a common base-class pays enough dividends right now. 

That leaves the startup sequence confusions, which I think we should resolve as follows:

- ```start()``` is a bad name for this event. I propose that we rename it to ```load()``` to give users a better indication of its function. The startup sequence will then be ```load``` -> ```configure``` -> ```running```. 
- The ```load``` event takes a Loader() object as argument, which exposes two methods ```add_addon``` and ```add_option```. This removes the implicit return and prevents users from accessing options during the ```load`` event. There is some ambiguity with the ```add_addon``` method proposed here, because the current structure replaces the running addon with the returned addon. We can consider this when we do the work - it may be useful to expand the Loader API to allow addons to load multiple sub-addons and to control their own removal from the addon mechanism separately.

**Comment by mhils** [comment]
> While the idea of deriving from a common base class is attractive, there are some issues. One is that our ad-hoc scripts at the moment are simply handlers at the module level. If we derive from a common class for addons, we would have two distinct types of entity here that will inevitably have different semantics if the base class does anything interesting. On balance, I don't think that moving to a common base-class pays enough dividends right now, so I propose we keep this as it is.

Generally 👍, I'd still welcome a base class that does nothing interesting but just defines function signatures and adds documentation. mypy would like it.

--------

> The startup event sequence for addons is quite confusing - we have start() -> configure() -> running(), all of which are called in different contexts, and have names that are close to each other semantically.

> The start() method takes an options argument, but addons can only create new options at this point, not read and use option values. This is confusing.

AFAIK we currently use `start` for three purposes:

1. Print "start" log entries etc.
2. Registering options
3. Returning an addon from a script (already outlined as problematic above)

Now, if we would make options declarative, they could be parsed before the initial `configure` and we are left with use cases 1 and 3. I believe 1 can just be moved into what currently is `running` in all cases. Last, we'll surely find another way to bootstrap into addons from scripts. For example, we could say that a script that defines a variable named `addon` will be replaced by this.

TL;DR: Ditch `start` entirely and make options declarative:
```python
class FooAddon:
    options = [
        mitmproxy.options.Option("foo", bool, True, "More foo, less bar.")
    ]
```

------

> The load event takes a Loader() object as argument, which exposes two methods add_addon and add_option

How would `add_addon` work with e.g. script reloading? Do you want to track addon dependencies to then unload multiple addons when unloading a script? I think `add_addon` is quite problematic in that regard, `add_option` probably not so much.

-----

> In keeping with our lightweight structure, I propose that addons declare custom event handlers simply by adding methods named event_NAME. We could use a decorator to mark events, but this is slightly out of step with the structure we have. I feel a decorator would be the better option if we do decide addons should derive from a common class.

Do we have that much of a pressing need for generic custom events? You may need to expand a bit on the use case. My gut feeling is that custom events would introduce a lot of complexity with actually very little benefit:

- From a user's perspective, why are some events prefixed with `event_` while others aren't? I don't want to teach the "oh we have two different classes of events, they are similar but not exactly equal" thing to users.
- What do we do with events where we have no receiver? Silently swallow? Raise?
- How do differentiate regular from custom events? Does that mean we have a separate `trigger` call for those?

Long story short, I feel it's a lot of pain and potential to fuck up for very few use cases. `master.addons["readfile"].read_file(path)` is a bit clumsy, but it's rather easy to grok what it does and we shouldn't need stuff like this too much.

**Comment by cortesi** [comment]
It would be handy if we could keep discussion here in the form of competing concrete proposals. To take your points in order:

- I've thought about this quite a bit, and my sense is that declarative options are not quite as good as an .add_option() call. Everything an addon does at the moment is event driven - this breaks that regularity. I feel this would tip us over to the point where addons should derive from a common class, and I'm not sure that step is warranted yet. I also think having a method here is more flexible, and allows us to modify the API more easily as this evolves.
- We won't have trouble handling add_addon for script reloading - addons can already boot into other addons. That's a different discussion, though. For the moment, let's consider a boot_into() method that is semantically equivalent to returning an addon from start() as it stands. The point is that I feel that's too under-specified, and I'd prefer an explicit all.
- I strongly feel that custom events are motivated. There are downsides to calling a method directly, for instance, we really want all actions an addon takes to occur within the handler context, so that it can consistently use logs and access master, as for events and options changes. A custom event structure lets us do this consistently. This also opens the way for addons to expose events that can, for instance, be bound to keyboard shortcuts down the track. In addition, we also have a number of internal places where a custom event is a nice solution, for instance the "processing_done" event we use to handle KeepServing in a general way. We've already discussed in Slack how I think we should handle the distinction between built-in events and custom events from an API perspective - I'll amend my proposal to outline this.

**Comment by cortesi** [comment]
I've been playing with this, and I've upated my proposal so that custom events are added in ```load``` by calling an ```.add_event``` method analogous to ```.add_option```.

**Comment by cortesi** [comment]
This is now largely in tree.
